Implement edit and remove for Clusters and Storage

jwmatthews opened the pull request:
I have 'remove' working.
Looking into 'edit' next.

@ibolton336 @eriknelson welcome any suggestions to improve

jwmatthews commented:
Part of the functionality for updating Cluster url and Secret value is working towards k8s.
Working on having the updates reflected in the redux Store, that is not yet working.
Also saw I broke add storage, so will need to fix.

jwmatthews commented:
At this point, edit and remove of Clusters is working.

Starting to look into 'Storage', first issue I see is that the storage secret was being sent to 'openshift-config' yet we declared it for 'mig' namespace.  I assume that the storage secret should be in 'mig' namespace so updated it to be there.

Have not addressed remove/edit of Storage yet, next on list to handle.

jwmatthews commented:
Have the below working now
 - Edit/Remove of Clusters
 - Edit/Remove of Storage

jwmatthews commented on src/app/home/components/DataList/Clusters/ClusterItem.tsx:
```diff
@@ -45,10 +52,23 @@ const ClusterItem = ({ cluster, clusterIndex, isLoading, migMeta, ...props }) =>
-                  <Button variant="danger">Remove</Button>
+                    <Button onClick={() => {
+                      // TODO:  Not sure best way to ensure that user wants to really delete this item.
+                      if (confirm('Are you sure?')) {
```
This is kind of ugly popping up an 'alert'.

Assuming there is a better approach to making this cleaner.

ibolton336 replied:
I think an alert popup is pretty standard although I am not seeing a PF4 component for one unfortunately. That may just mean we need to just style the popup to look like pf4. 

something like this https://react.semantic-ui.com/addons/confirm/#types-confirm

andyvesel replied:
Maybe we can use some PF4 modal?

ibolton336 replied:
@jwmatthews added a pf4 modal component for this. going to merge in to your branch .

jwmatthews commented on src/app/storage/duck/operations.ts:
```diff
@@ -66,6 +67,42 @@ const addStorage = storageValues => {
+        client.patch(migStorageResource, storageValues.name, migStorage),
+      ]);
+
+      dispatch(fetchStorage());
```
Wasn't sure if we are supposed to issue an updatedSuccess to redux...could use some help knowing what the proper thing to do is around here.

ibolton336 replied:
Yeah we have been sending success actions at the end of successful operations/api calls. They normally have just changed loading status & we havent implemented a success alert system.  We should open an issue to add  success notifications that auto dismiss after a set time.

jwmatthews replied:
I see @andyvesel wrote up an issue to capture the need for success alert system:  https://github.com/fusor/mig-ui/issues/214

jwmatthews replied:
Addressed.

jwmatthews commented on src/app/home/components/DataList/Storage/StorageItem.tsx:
```diff
@@ -39,10 +49,24 @@ const StorageItem = ({ storage, storageIndex, isLoading, ...props }) => {
-                  <Button variant="danger">Remove</Button>
+                  <Button onClick={() => {
+                      // TODO:  Not sure best way to ensure that user wants to really delete this item.
+                      if (confirm('Are you sure?')) {
```
Again, looking for better way to confirm before delete.

jwmatthews commented on src/app/storage/duck/operations.ts:
```diff
@@ -76,18 +113,28 @@ function checkConnection() {
+        client.delete(migStorageResource, id),
+      ]);
+
+      dispatch(fetchStorage());
```
Should we issue some sort of remove success?

jwmatthews commented:
@ibolton336 @eriknelson @andyvesel 

To get this ready to merge, I know of 2 areas to address, please let me know if there are others to improve:
1. Add a 'Confirm' component, perhaps something like https://react.semantic-ui.com/addons/confirm/#types-confirm, with PF4 components
1. Ensure after performing the update/remove actions for both Cluster and Storage we are sending 'success' events.

Any other comments to improve?  I am new to React/Redux, appreciate any help to learn and get better.

andyvesel commented:
@jwmatthews in addition to ^ there's also an issue: after update/delete cluster, I always need to refresh the page manually to see that cluster was actually updated/deleted.

jwmatthews commented on src/app/cluster/duck/operations.ts:
```diff
@@ -76,19 +79,67 @@ const addCluster = clusterValues => {
+        client.patch(clusterRegResource, clusterValues.name, clusterReg),
+        client.patch(secretResource, clusterValues.name, tokenSecret),
+      ]);
+      dispatch(fetchClusters());
```
Assume I need to come back to this to add a update success message

jwmatthews replied:
Addressed

jwmatthews commented on src/app/cluster/duck/operations.ts:
```diff
@@ -76,19 +79,67 @@ const addCluster = clusterValues => {
+        client.delete(migClusterResource, id),
+      ]);
+
+      dispatch(fetchClusters());
```
Assume need a remove cluster success message

jwmatthews replied:
Addressed

jwmatthews commented:
I think last thing left is for me to update the PR with some success messages when edit/remove has succeeded or messages when they've failed.

ibolton336 commented:
> I think last thing left is for me to update the PR with some success messages when edit/remove has succeeded or messages when they've failed.

Yep that should be it.

jwmatthews commented:
As I'm testing I'm seeing "redux-logger.js?6f8a:1 Uncaught (in promise) TypeError: Cannot read property 'type' of undefined" after editing a cluster.

Need to debug further.

jwmatthews commented on src/app/cluster/duck/operations.ts:
```diff
@@ -76,19 +79,70 @@ const addCluster = clusterValues => {
+        client.patch(clusterRegResource, clusterValues.name, clusterReg),
+        client.patch(secretResource, clusterValues.name, tokenSecret),
+      ]);
+      dispatch(updateClusterSuccess);
```
@ibolton336 please take a look to see if I've updating this as you expected....I feel like I'm not doing this correctly.

I haven't created any function to handle the update/remove success.... for either clusters or storage, not sure if there is some magic creating this or if I need to define it.

jwmatthews replied:
Addressed

jwmatthews commented:
I think I got the redux pieces correct now.
I wasn't adjusting the actions/reducers so there was no change to local state.   I think last commit addresses this.

jwmatthews commented:
Rebased from master, now seeing below

ERROR in /git/fusor/mig-ui/src/app/common/components/ConfirmModal.tsx
./src/app/common/components/ConfirmModal.tsx
[tsl] ERROR in /git/fusor/mig-ui/src/app/common/components/ConfirmModal.tsx(20,7)
      TS2322: Type '(isConfirmed: any) => void' is not assignable to type '() => void'.

eriknelson commented:
> ERROR in /git/fusor/mig-ui/src/app/common/components/ConfirmModal.tsx
./src/app/common/components/ConfirmModal.tsx
[tsl] ERROR in /git/fusor/mig-ui/src/app/common/components/ConfirmModal.tsx(20,7)
TS2322: Type '(isConfirmed: any) => void' is not assignable to type '() => void'.

Looks like PF may have made a breaking change to their API so the `onClose` prop expects a `() => void` instead of `(isConfirmed: any) => void`?

jwmatthews commented on src/client/resources/conversions.ts:
```diff
@@ -116,7 +159,7 @@ export function createStorageSecret(
     data: {
       'aws-access-key-id': encodedAccessKey,
-      'aws-secret-access-key': encodedSecretKey,
+      'aws-secret-access-key-id': encodedSecretKey,
```
Need to look more to determine which value this should be.

'aws-secret-access-key' or 'aws-secret-access-key-id'.

jwmatthews replied:
Addressed.

jwmatthews commented:
Getting close to OK to merge.

Last issue is that AddStorage is not handling the aws secret access key consistently.
Need to check what the key should be:  'aws-secret-access-key' or 'aws-secret-access-key-id'.

Will update PR to fix that, then think it will be ready.

jwmatthews commented:
Fixed all issues I know about.

Ready for review and merge if OK.

ibolton336 commented on src/app/cluster/duck/operations.ts:
```diff
@@ -77,19 +80,79 @@ const addCluster = clusterValues => {
+        client.get(migClusterResource, clusterValues.name)
+      ]);
+      const cluster = arr.reduce((accum, res) => {
+        accum[res.data.kind] = res.data;
```
In local demo mode, this piece is failing. Would be good to add a TODO comment here to add update support for local development( if that is something that we plan on doing. not sure there is much left to develop on this so could be a waste of time. )

jwmatthews replied:
Planning to work on local support with Andy.  I think local is all broken right now.

ibolton336 commented on src/app/cluster/duck/reducers.ts:
```diff
@@ -34,7 +34,16 @@ export const addClusterSuccess = (state = INITIAL_STATE, action) => {
-  return { ...state };
+  return {
+    ...state,
+    clusterList: state.clusterList.filter(item => item.Cluster.metadata.name !== action.name),
```
We been using the MigCluster object (not Cluster) to reference the metadata.name value. This is a nitpick but we should be consistent with how we are accessing this data.

jwmatthews replied:
Will update.

eriknelson commented:
> *@jwmatthews* commented on this pull request.
> ------------------------------
>
> In src/client/resources/conversions.ts
> <https://github.com/fusor/mig-ui/pull/211#discussion_r291757193>:
>
> > @@ -116,7 +159,7 @@ export function createStorageSecret(
>      apiVersion: 'v1',
>      data: {
>        'aws-access-key-id': encodedAccessKey,
> -      'aws-secret-access-key': encodedSecretKey,
> +      'aws-secret-access-key-id': encodedSecretKey,
>
> Need to look more to determine which value this should be.
>
> 'aws-secret-access-key' or 'aws-secret-access-key-id'.
>
This should be 'aws-secret-access-key` without the `-id` suffix. Was a bug
I had originally introduced, realized it was a bad key during integration
testing last week and updated.

jwmatthews commented:
> ***@***.**** commented on this pull request. ------------------------------ In src/client/resources/conversions.ts <[#211 (comment)](https://github.com/fusor/mig-ui/pull/211#discussion_r291757193)>: > @@ -116,7 +159,7 @@ export function createStorageSecret( apiVersion: 'v1', data: { 'aws-access-key-id': encodedAccessKey, - 'aws-secret-access-key': encodedSecretKey, + 'aws-secret-access-key-id': encodedSecretKey, Need to look more to determine which value this should be. 'aws-secret-access-key' or 'aws-secret-access-key-id'.
> This should be 'aws-secret-access-key` without the `-id` suffix. Was a bug I had originally introduced, realized it was a bad key during integration testing last week and updated.

ACK

That was an older comment, since then I looked at the mig-cluster types and confirmed 'aws-secret-access-key' and updated PR.

Think we are good now and on same page of using 'aws-secret-access-key'.
